Library/Homebrew/formula.rb

```diff
@@ -1505,7 +1505,12 @@ def recursive_requirements(&block)
   # Returns a list of Dependency objects that are required at runtime.
   # @private
   def runtime_dependencies
-    recursive_dependencies.reject(&:build?)
+    recursive_dependencies do |_dependent, dependency|
+      Dependency.prune if dependency.build?
+      if dependency.optional? || dependency.recommended?
```

Comment:
`if build.without?(dependency) && (dependency.optional? || dependency.recommended?)` may be easier to follow. Could also consider having `optional_dependency = dependency.optional? || dependency.recommended?` before that, too.